File: apps/backend/domain/attendance.go
package domain

import (
	"time"
)

type AttendanceStatus string

const (
	OnTime     AttendanceStatus = "on_time"
	Late       AttendanceStatus = "late"
	EarlyLeave AttendanceStatus = "early_leave"
	Absent     AttendanceStatus = "absent"
)

type Attendance struct {
	ID           uint             `gorm:"primaryKey"`
	EmployeeID   uint             `gorm:"not null"`
	Employee     Employee         `gorm:"foreignKey:EmployeeID"`
	Date         time.Time        `gorm:"type:date;not null"`
	ClockIn      *time.Time       `gorm:"type:timestamp"`
	ClockOut     *time.Time       `gorm:"type:timestamp"`
	ClockInLat   *float64         `gorm:"type:float"`
	ClockInLong  *float64         `gorm:"type:float"`
	ClockOutLat  *float64         `gorm:"type:float"`
	ClockOutLong *float64         `gorm:"type:float"`
	WorkHours    *float64         `gorm:"type:float"`
	Status       AttendanceStatus `gorm:"type:attendance_status;not null;default:on_time"`

	CreatedAt time.Time `gorm:"autoCreateTime"`
	UpdatedAt time.Time `gorm:"autoUpdateTime"`
}

func (a *Attendance) TableName() string {
	return "attendances"
}

File: apps/frontend/src/types/attendance.ts
import { Employee } from "./employee";
import { WorkSchedule } from "./work-schedule.types";

export interface Attendance {
	id: number;
	employee_id: number;
	employee?: Employee;
	work_schedule_id: number;
	work_schedule?: WorkSchedule;
	date: string;
	clock_in: string | null;
	clock_out: string | null;
	clock_in_lat: number | null;
	clock_in_long: number | null;
	clock_out_lat: number | null;
	clock_out_long: number | null;
	work_hours: number | null;
	status: string;
	created_at: string;
	updated_at: string;
}

export interface ClockInAttendanceRequest {
	employee_id: number;
	work_schedule_id: number;
	date?: string;
	clock_in?: string;
	clock_in_lat: number;
	clock_in_long: number;
}

export interface ClockOutAttendanceRequest {
	employee_id: number;
	date?: string;
	clock_out?: string;
	clock_out_lat: number;
	clock_out_long: number;
}

export interface AttendanceFormData {
	attendance_type: "clock-in" | "clock-out";
	clock_in_request?: ClockInAttendanceRequest;
	clock_out_request?: ClockOutAttendanceRequest;
}

// Backend response format
export interface AttendanceApiResponse<T> {
	status: number;
	message: string;
	data?: T;
	error?: string;
}

export interface AttendanceListResponse {
	items: Attendance[];
	pagination: {
		page: number;
		size: number;
		total: number;
		total_pages: number;
	};
}
